python/ray/dag/compiled_dag_node.py

```diff
@@ -572,180 +631,124 @@ def prepare(self, overlap_gpu_communication: bool = False):
                         nccl_group_id
                     )
                     assert nccl_group is not None
-                    if not isinstance(self._recv_stream, nullcontext):
-                        assert self._recv_stream == nccl_group.recv_stream, (
+                    if not isinstance(self.stream, nullcontext):
+                        assert self.stream == nccl_group.recv_stream, (
                             "Currently all torch tensor input channels of a "
                             "Compiled Graph task should use the same recv cuda stream."
                         )
-                    self._recv_stream = nccl_group.recv_stream
-
-    def wrap_and_set_intermediate_future(
-        self, val: Any, wrap_in_gpu_future: bool
-    ) -> None:
-        """
-        Wrap the value in a `DAGOperationFuture` and store to the intermediate future.
-        The value corresponds to result of a READ or COMPUTE operation.
-
-        If wrap_in_gpu_future is True, the value will be wrapped in a GPUFuture,
-        Otherwise, the future will be a ResolvedFuture.
-
-        Args:
-            val: The value to wrap in a future.
-            wrap_in_gpu_future: Whether to wrap the value in a GPUFuture.
-        """
-        assert self._intermediate_future is None
-
-        if wrap_in_gpu_future:
-            future = GPUFuture(val)
-        else:
-            future = ResolvedFuture(val)
-        self._intermediate_future = future
-
-    def reset_and_wait_intermediate_future(self) -> Any:
-        """
-        Reset the intermediate future and wait for the result.
+                    self.stream = nccl_group.recv_stream
 
-        The wait does not block the CPU because:
-        - If the future is a ResolvedFuture, the result is immediately returned.
-        - If the future is a GPUFuture, the result is only waited by the current
-            CUDA stream, and the CPU is not blocked.
-
-        Returns:
-            The result of a READ or COMPUTE operation from the intermediate future.
-        """
-        future = self._intermediate_future
-        self._intermediate_future = None
-        return future.wait()
+        if self.requires_nccl_collective:
+            assert isinstance(self.nccl_op, _CollectiveOperation)
+            self.stream = self.nccl_op.get_communicator().coll_stream
 
-    def _read(self, overlap_gpu_communication: bool) -> bool:
+    def exec_operation_with_contexts(
```

Comment:
This naming is slightly confusing because it sounds like `context` should be passed as an argument to this function. Maybe just `exec_operation` and `_exec_operation_without_context` for the inner method.